### PR TITLE
Support for Arbitrarily Complex Predicates

### PR DESCRIPTION
**Pull request by Bren2010**
Integrated current code with [Bren2010/MSP](https://github.com/Bren2010/msp).  Rather than passing in an `Owners` or `Left/RightOwners` array with encryption requests, this allows you to pass in a `Predicate` key with your predicate over known users.

I believe it's backwards-compatible and that there are no new bugs.

### CODE REVIEW

**Comment by kisom** [comment]
I haven't gotten a chance to read much of the diff, but I did notice it brings in an external dependency. I'd prefer it if `msp` was vendored into the package (I did notice it doesn't have external dependencies itself).

**Comment by Bren2010** [comment]
No problem at all.  :)  Does the last commit do what you want?  It can be added as a submodule as well.

**Comment by kisom** [comment]
It does --- thanks!

The idea here is that we can always build a `redoctober` server from this repository without having to rely on external code, as those have a habit of changing or disappearing. A submodule would still keep an external repository as a dependency, so this is actually the ideal solution.

**Comment by ejcx** [comment]
Got some errors deep in the msp library when trying this with some basic predicates. Not sure if I did something wrong. Here's an example.

```
goroutine 5 [running]:
github.com/cloudflare/redoctober/msp.StringToRaw(0x111841fa, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /Users/e/go/src/github.com/cloudflare/redoctober/msp/raw.go:136 +0xab8
github.com/cloudflare/redoctober/msp.StringToMSP(0x11185560, 0x19, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /Users/e/go/src/github.com/cloudflare/redoctober/msp/msp.go:115 +0xa1
github.com/cloudflare/redoctober/cryptor.(*EncryptedData).wrapKey(0x122880c0, 0x39e2c0, 0x1221c020, 0x10, 0x10, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/e/go/src/github.com/cloudflare/redoctober/cryptor/cryptor.go:343 +0xd9f
github.com/cloudflare/redoctober/cryptor.(*Cryptor).Encrypt(0x39e018, 0x1221de24, 0xc, 0xc, 0x11185580, 0x1, 0x4, 0x0, 0x0, 0x0, ...)
        /Users/e/go/src/github.com/cloudflare/redoctober/cryptor/cryptor.go:477 +0x1c7
github.com/cloudflare/redoctober/core.Encrypt(0x111be200, 0x81, 0x200, 0x0, 0x0, 0x0, 0x0, 0x0)
        /Users/e/go/src/github.com/cloudflare/redoctober/core/core.go:436 +0x312
main.func·003()
        /Users/e/redoctober/redoctober.go:229 +0xbb
created by main.main
        /Users/e/redoctober/redoctober.go:246 +0x2c0
```

Here's the curl request I sent it

```
curl 'https://localhost:8080/encrypt' -H 'Origin: https://localhost:8080' -H 'Accept-Encoding: gzip, deflate' -H 'Accept-Language: en-US,en;q=0.8' -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/44.0.2403.125 Safari/537.36' -H 'Content-Type: application/x-www-form-urlencoded; charset=UTF-8' -H 'Accept: application/json, text/javascript, */*; q=0.01' -H 'Referer: https://localhost:8080/' -H 'X-Requested-With: XMLHttpRequest' -H 'Connection: keep-alive' --data '{"Name":"evan","Password":"evan","Minimum":2,"Predicate":"((evan or evan1) and evan2)","Labels":["pred"],"Data":"YXNkZmFzZGZhc2Rm"}' --compressed -k
```

I also tried a variation without the outer paren.

**Comment by Bren2010** [comment]
@ejcx If you change `((evan or evan1) and evan2)` to `((evan | evan1) & evan2)`, does that fix it?

**Comment by ejcx** [comment]
Ah...duh. I didn't read enough of the docs. Thank you

**Comment by ejcx** [comment]
Hey Bren,
I'm trying to make my branch work with your branch, so I've been going over a bunch of your code. Would you mind fixing the issue in msp throwing errors when you put in bad syntax like I did

Example:  `((evan1 or evan2) and evan)`

It should just return an error if it sees something it shouldn't.

**Comment by Bren2010** [comment]
@ejcx I can't replicate your panic (I think it's a panic?).  For the command you posted, `curl` returns
```
{"Status":"Stacks are invalid size."}
```
and the server logs an error
```
Error encrypting: Stacks are invalid size.
```

Which... looking at now, is not a helpful error message and I will fix that.  But what code are you running that panics?  (I tried to steal ejxc/redoctober, but I get build errors.)

**Comment by kisom** [comment]
FWIW, I ran into this as well using the predicate "(alice or bob and
carol)". The following Python code will replicate this:

```python
import requests
import json

# Replace with your credentials.
user = 'alice'
password = 'password'
request_data = {

        'Name': user,
        'Password': password,
        'Predicate': '(alice or bob and carol)',
        'Data': 'hello, world'.encode('base64')
}

request = requests.post('https://127.0.0.1:8080/encrypt',
                        verify=False,
                        data=json.dumps(request_data))
print request.content
```

**Comment by ejcx** [comment]
Hey Bren,
Just to update you I fixed whatever issue it was that I was having. Doing a bit of work to merge your PR with my branch.
Evan

**Comment by grittygrease** [comment]
@Bren2010, can you rebase this on master and update the UI's "Encrypt Data" portion to include an input field for your predicate string as an alternative to "Owners"?

**Comment by grittygrease** [comment]
It would also be helpful to add an example in the README

**Comment by grittygrease** [comment]
Another thing about the UI, when attempting to decrypt a secret that does not have enough delegations, it presents two red text fields:
"Not enough shares to recover."
and
"SyntaxError: Unexpected token <"

The second doesn't need to be printed.

**Comment by grittygrease** [comment]
I think the output can also be improved. The output of a decrypt is:
{"Data":"aGVsbG8K","Secure":true,"Delegates":["Shares"]}
where in the regular case it's:
{"Data":"aGVsbG8K","Secure":true,"Delegates":["a","b"]}

The caller doesn't know which delegations were used in the first case.

**Comment by grittygrease** [comment]
I also hit a panic while testing:

panic: crypto/aes: input not full block

goroutine 5 [running]:
crypto/aes.(*aesCipher).Encrypt(0xc2080b2090, 0xc208090000, 0x10, 0x10, 0xc208090251, 0xf, 0xf)
	/usr/local/go/src/crypto/aes/cipher.go:50 +0x77
github.com/cloudflare/redoctober/cryptor.(*EncryptedData).wrapKey(0xc208096790, 0x48fd20, 0xc2080901c0, 0x10, 0x10, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/nick/code/gospace/src/github.com/cloudflare/redoctober/cryptor/cryptor.go:357 +0x1620
github.com/cloudflare/redoctober/cryptor.(*Cryptor).Encrypt(0x48f390, 0xc208090176, 0x6, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/nick/code/gospace/src/github.com/cloudflare/redoctober/cryptor/cryptor.go:471 +0x224
github.com/cloudflare/redoctober/core.Encrypt(0xc208118200, 0x51, 0x200, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/nick/code/gospace/src/github.com/cloudflare/redoctober/core/core.go:290 +0x407
main.func·003()
	/Users/nick/code/gospace/src/github.com/cloudflare/redoctober/redoctober.go:222 +0xf7
created by main.main
	/Users/nick/code/gospace/src/github.com/cloudflare/redoctober/redoctober.go:239 +0x3be

**Comment by Bren2010** [comment]
@grittygrease 
- Rebased
- 9e594e2 fixes your AES panic and the casting panics from before.
- 9bfde59 adds the delegation list
- 6ed973e are the UI changes

>  "SyntaxError: Unexpected token <"

I don't see this error?

**Comment by grittygrease** [comment]
The "Get Owners" API should be updated to support returning the owner predicate.

**Comment by grittygrease** [comment]
"SyntaxError: Unexpected token <" must have been fixed in the rebase.

File: msp/msp.go [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+type TraceSlice []TraceElem
+
+func (ts TraceSlice) Len() int      { return len(ts) }
+func (ts TraceSlice) Swap(i, j int) { ts[i], ts[j] = ts[j], ts[i] }
+
+func (ts TraceSlice) Less(i, j int) bool {
+	return len(ts[i].trace) < len(ts[j].trace)
+}
+
+func (ts *TraceSlice) Push(te interface{}) { *ts = append(*ts, te.(TraceElem)) }
```
Comment (lziest) [inline msp/msp.go]: why interface?

Comment (Bren2010) [inline msp/msp.go]: TraceSlice implements heap.Interface and sort.Interface.  Push and Pop are from heap.Interface.

File: msp/msp.go [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+			name := cond.(String).string
+			if _, ok := out[name]; ok {
+				out[name] = append(out[name], m.encode(share, modulus))
+			} else if (*db).ValidUser(name) {
+				out[name] = [][]byte{m.encode(share, modulus)}
+			} else {
+				return out, errors.New("Unknown user in predicate.")
+			}
+
+		default:
```
Comment (lziest) [inline msp/msp.go]: switch on type MSP or Formatted as other switches?

File: msp/msp.go [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+	GetShare(string) ([][]byte, error)
+}
+
+type Condition interface { // Represents one condition in a predicate
+	Ok(*UserDatabase) bool
+}
+
+type String struct { // Type of condition
+	string
+	index int
```
Comment (lziest) [inline msp/msp.go]: I haven't gotten the full understanding of the usage of 'index'

Comment (Bren2010) [inline msp/msp.go]: If you're reading a formatted or raw parse tree from left to right, the index is the number of times you've seen this user so far.  When we're recovering the secret, if we want to satisfy this node, we'll use the `index`th share of user `string`. (A user will have `n` shares if their name appears in the predicate `n` times.)

File: keycache/keycache.go [hunk 4]
```diff
@@ -100,9 +100,9 @@ func (cache *Cache) Valid(name, user string, labels []string) (present bool) {
 	return
 }
 
-// matchUser returns the matching active user if present
+// MatchUser returns the matching active user if present
 // and a boolean to indicate its presence.
-func (cache *Cache) matchUser(name, user string, labels []string) (out ActiveUser, present bool) {
+func (cache *Cache) MatchUser(name, user string, labels []string) (out ActiveUser, present bool) {
```
Comment (lziest) [inline keycache/keycache.go]: probably need to rebase and there would be needs to addressed conflicts.

File: msp/msp.go (hunk 3 above)
Comment (lziest) [inline msp/msp.go]: Thanks! That explains a lot. Can we think of a better name than "String"? "User"?

File: msp/msp.go [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+	"math/big"
+	"strings"
+)
+
+// A UserDatabase is an abstraction over the name -> share map returned by the
+// secret splitter that allows an application to only decrypt or request shares
+// when needed, rather than re-build a partial map of known data.
+type UserDatabase interface {
+	ValidUser(name string) bool
+	CanGetShare(string) bool
```
Comment (lziest) [inline msp/msp.go]: can we say CanGetShare(name string)? or "string" is actually a predicate? I like the variable to be named.

File: msp/msp.go [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+	GetShare(string) ([][]byte, error)
```
Comment (lziest) [inline msp/msp.go]: ditto.

File: msp/msp.go [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+		modulus.Add(modulus, big.NewInt(0).Lsh(big.NewInt(1), 96))
+		modulus.Sub(modulus, big.NewInt(1))
+
+	case 224:
+		modulus = big.NewInt(1) // 2^224 - 2^96 + 1
+		modulus.Lsh(modulus, 224)
+		modulus.Sub(modulus, big.NewInt(0).Lsh(big.NewInt(1), 96))
+		modulus.Add(modulus, big.NewInt(1))
+
+	default: // Silent fail.
```
Comment (lziest) [inline msp/msp.go]: I guess there is no fail here? just default to 2^127 - 1?

Comment (lziest) [inline msp/msp.go]: Actually I don't mind we panic here if the value is not 256, 224, or 127. :-D

File: cryptor/cryptor.go [hunk 8]
```diff
@@ -259,8 +302,6 @@ func (encrypted *EncryptedData) wrapKey(records *passvault.Records, clearKey []b
 				encrypted.KeySet = append(encrypted.KeySet, out)
 			}
 		}
-
-		return nil
```
Comment (jkroll-cf) [inline cryptor/cryptor.go]: The comment at the top of this function is no longer accurate. Please fix it.

**Comment by grittygrease** [comment]
If there are no more comments, I'm going to merge this.

**Pull request merged**